src/app/icd/client/ICDClientInfoPersistentStorage.h

```diff
@@ -0,0 +1,95 @@
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/ScopedNodeId.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <lib/support/StorageKeyName.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * The ICDClientInfoPersistentStorage class is an abstract interface that defines the operations
+ * for storing, retrieving and deleting ICD client information in persistent storage.
+ */
+class ICDClientInfoPersistentStorage
+{
+public:
+    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
+
+    virtual ~ICDClientInfoPersistentStorage() = default;
+
+    /**
+     * Iterate through persisted ICD Client Info
+     *
+     * @return A valid iterator on success. Use CommonIterator accessor to retrieve ICDClientInfo
+     */
+    virtual ICDClientInfoIterator * IterateICDClientInfo() = 0;
+
+    /**
+     * Called during ICD device registration in commissioning, sdk provides the clientInfo and application
+     * provides raw key data, the shared key handle in clientInfo is updated based upon raw key data
+     *
+     * @param[inout] aICDClientInfo the ICD Client information to be updated with keyData and be saved
+     * @param[in] aKeyData raw key data provided by application
+     */
+    virtual CHIP_ERROR SetKey(ICDClientInfo & aClientInfo, const ByteSpan aKeyData) = 0;
+
+    /**
+     * Store updated ICD ClientInfo to storage when ICD registration completes or check-in message
+     * comes.
+     *
+     * @param[in] aICDClientInfo the updated ICD Client Info.
+     */
+    virtual CHIP_ERROR StoreICDInfoEntry(ICDClientInfo & aICDClientInfo) = 0;
+    /**
+     * Delete ICD Client persistent information associated with the specified scope node Id.
+     * when ICD device is unparied/removed, the correponding entry in ICD persistant storage is removed.
+     * @param aPeerNodeId scope node with peer node id and fabric index
+     */
+    virtual CHIP_ERROR DeleteEntry(ScopedNodeId aPeerNodeId) = 0;
+
+    /**
+     * Remove all ICDClient persistent informations associated with the specified
+     * fabric index.  If no entries for the fabric index exist, this is a no-op
+     * and is considered successful.
+     * When the whole fabric is removed, all entries from persistent storage in current fabric index is removed.
+     *
+     * @param[in] fabricIndex the index of the fabric for which to remove ICDClient persistent information
+     */
+    virtual CHIP_ERROR DeleteAllEntries(FabricIndex aFabricIndex) = 0;
+
+    /**
+     * Validate received ICD Check-in message payload, return true when the key decription is successful
+     * Consumer has to provide keys and see whether it's the right key, then return the correponding ICDClientInfo
+     * @param[in] aPayload received checkIn Message payload
+     * @param[out] ICDClientInfo matched clientInfo
+     */
+    virtual bool ValidateKey(const ByteSpan & aPayload, ICDClientInfo & aClientInfo) = 0;
```

Comment:
Does this handle repeated counters?

Reply:
yes, it would validate the counter as well, so yes, repeated counters..

Reply:
updated doc and rename the API to ValidateCheckInPayload. thanks